{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5b5ee5c4",
   "metadata": {},
   "source": [
    "This can read from a locally-staged hail table directory, or use one in GCS. The table must have a pre-computed `.info.VRS` field. This is computed using `tgg_methods` `vrs_annotation_batch.py`, which uses the `vrs-python` `vcf_annotation.py`\n",
    "\n",
    "https://github.com/broadinstitute/tgg_methods/blob/master/vrs/vrs_annotation_batch.py (last checked at `a0002f02fbd5dd25487b261e94081a3daec29c64`)\n",
    "\n",
    "This uses gnomad_methods functions `gnomad_gks` and `add_gks_vrs`, and `add_gks_va`, to return GKS structures for hail tables that are in the gnomad v3 schema, off the branch in this pull request: https://github.com/broadinstitute/gnomad_methods/pull/556\n",
    "\n",
    "The JSON schema is in `ga4gh/va-spec` on the `gk-pilot` branch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "afc240ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       "        .bk-notebook-logo {\n",
       "            display: block;\n",
       "            width: 20px;\n",
       "            height: 20px;\n",
       "            background-image: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAUCAYAAACNiR0NAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAABx0RVh0U29mdHdhcmUAQWRvYmUgRmlyZXdvcmtzIENTNui8sowAAAOkSURBVDiNjZRtaJVlGMd/1/08zzln5zjP1LWcU9N0NkN8m2CYjpgQYQXqSs0I84OLIC0hkEKoPtiH3gmKoiJDU7QpLgoLjLIQCpEsNJ1vqUOdO7ppbuec5+V+rj4ctwzd8IIbbi6u+8f1539dt3A78eXC7QizUF7gyV1fD1Yqg4JWz84yffhm0qkFqBogB9rM8tZdtwVsPUhWhGcFJngGeWrPzHm5oaMmkfEg1usvLFyc8jLRqDOMru7AyC8saQr7GG7f5fvDeH7Ej8CM66nIF+8yngt6HWaKh7k49Soy9nXurCi1o3qUbS3zWfrYeQDTB/Qj6kX6Ybhw4B+bOYoLKCC9H3Nu/leUTZ1JdRWkkn2ldcCamzrcf47KKXdAJllSlxAOkRgyHsGC/zRday5Qld9DyoM4/q/rUoy/CXh3jzOu3bHUVZeU+DEn8FInkPBFlu3+nW3Nw0mk6vCDiWg8CeJaxEwuHS3+z5RgY+YBR6V1Z1nxSOfoaPa4LASWxxdNp+VWTk7+4vzaou8v8PN+xo+KY2xsw6une2frhw05CTYOmQvsEhjhWjn0bmXPjpE1+kplmmkP3suftwTubK9Vq22qKmrBhpY4jvd5afdRA3wGjFAgcnTK2s4hY0/GPNIb0nErGMCRxWOOX64Z8RAC4oCXdklmEvcL8o0BfkNK4lUg9HTl+oPlQxdNo3Mg4Nv175e/1LDGzZen30MEjRUtmXSfiTVu1kK8W4txyV6BMKlbgk3lMwYCiusNy9fVfvvwMxv8Ynl6vxoByANLTWplvuj/nF9m2+PDtt1eiHPBr1oIfhCChQMBw6Aw0UulqTKZdfVvfG7VcfIqLG9bcldL/+pdWTLxLUy8Qq38heUIjh4XlzZxzQm19lLFlr8vdQ97rjZVOLf8nclzckbcD4wxXMidpX30sFd37Fv/GtwwhzhxGVAprjbg0gCAEeIgwCZyTV2Z1REEW8O4py0wsjeloKoMr6iCY6dP92H6Vw/oTyICIthibxjm/DfN9lVz8IqtqKYLUXfoKVMVQVVJOElGjrnnUt9T9wbgp8AyYKaGlqingHZU/uG2NTZSVqwHQTWkx9hxjkpWDaCg6Ckj5qebgBVbT3V3NNXMSiWSDdGV3hrtzla7J+duwPOToIg42ChPQOQjspnSlp1V+Gjdged7+8UN5CRAV7a5EdFNwCjEaBR27b3W890TE7g24NAP/mMDXRWrGoFPQI9ls/MWO2dWFAar/xcOIImbbpA3zgAAAABJRU5ErkJggg==);\n",
       "        }\n",
       "    </style>\n",
       "    <div>\n",
       "        <a href=\"https://bokeh.org\" target=\"_blank\" class=\"bk-notebook-logo\"></a>\n",
       "        <span id=\"cbf3ce63-372a-42c1-bbfc-02327b3d1a36\">Loading BokehJS ...</span>\n",
       "    </div>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": "(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\nconst JS_MIME_TYPE = 'application/javascript';\n  const HTML_MIME_TYPE = 'text/html';\n  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n  const CLASS_NAME = 'output_bokeh rendered_html';\n\n  /**\n   * Render data to the DOM node\n   */\n  function render(props, node) {\n    const script = document.createElement(\"script\");\n    node.appendChild(script);\n  }\n\n  /**\n   * Handle when an output is cleared or removed\n   */\n  function handleClearOutput(event, handle) {\n    const cell = handle.cell;\n\n    const id = cell.output_area._bokeh_element_id;\n    const server_id = cell.output_area._bokeh_server_id;\n    // Clean up Bokeh references\n    if (id != null && id in Bokeh.index) {\n      Bokeh.index[id].model.document.clear();\n      delete Bokeh.index[id];\n    }\n\n    if (server_id !== undefined) {\n      // Clean up Bokeh references\n      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n      cell.notebook.kernel.execute(cmd_clean, {\n        iopub: {\n          output: function(msg) {\n            const id = msg.content.text.trim();\n            if (id in Bokeh.index) {\n              Bokeh.index[id].model.document.clear();\n              delete Bokeh.index[id];\n            }\n          }\n        }\n      });\n      // Destroy server and session\n      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n      cell.notebook.kernel.execute(cmd_destroy);\n    }\n  }\n\n  /**\n   * Handle when a new output is added\n   */\n  function handleAddOutput(event, handle) {\n    const output_area = handle.output_area;\n    const output = handle.output;\n\n    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n      return\n    }\n\n    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n\n    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n      // store reference to embed id on output_area\n      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n    }\n    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n      const bk_div = document.createElement(\"div\");\n      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n      const script_attrs = bk_div.children[0].attributes;\n      for (let i = 0; i < script_attrs.length; i++) {\n        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n      }\n      // store reference to server id on output_area\n      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n    }\n  }\n\n  function register_renderer(events, OutputArea) {\n\n    function append_mime(data, metadata, element) {\n      // create a DOM node to render to\n      const toinsert = this.create_output_subarea(\n        metadata,\n        CLASS_NAME,\n        EXEC_MIME_TYPE\n      );\n      this.keyboard_manager.register_events(toinsert);\n      // Render to node\n      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n      render(props, toinsert[toinsert.length - 1]);\n      element.append(toinsert);\n      return toinsert\n    }\n\n    /* Handle when an output is cleared or removed */\n    events.on('clear_output.CodeCell', handleClearOutput);\n    events.on('delete.Cell', handleClearOutput);\n\n    /* Handle when a new output is added */\n    events.on('output_added.OutputArea', handleAddOutput);\n\n    /**\n     * Register the mime type and append_mime function with output_area\n     */\n    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n      /* Is output safe? */\n      safe: true,\n      /* Index of renderer in `output_area.display_order` */\n      index: 0\n    });\n  }\n\n  // register the mime type if in Jupyter Notebook environment and previously unregistered\n  if (root.Jupyter !== undefined) {\n    const events = require('base/js/events');\n    const OutputArea = require('notebook/js/outputarea').OutputArea;\n\n    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n      register_renderer(events, OutputArea);\n    }\n  }\n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    const el = document.getElementById(\"cbf3ce63-372a-42c1-bbfc-02327b3d1a36\");\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.1.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.1.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.1.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.1.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.1.1.min.js\"];\n  const css_urls = [];\n\n  const inline_js = [    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\nfunction(Bokeh) {\n    }\n  ];\n\n  function run_inline_js() {\n    if (root.Bokeh !== undefined || force === true) {\n          for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\nif (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(\"cbf3ce63-372a-42c1-bbfc-02327b3d1a36\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));",
      "application/vnd.bokehjs_load.v0+json": ""
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import hail as hl\n",
    "import json\n",
    "from typing import List, Tuple\n",
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4e4985d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Initializing Hail with default parameters...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gs://gcp-public-data--gnomad/release/4.0/ht/exomes/gnomad.exomes.v4.0.sites.ht\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "SLF4J: No SLF4J providers were found.\n",
      "SLF4J: Defaulting to no-operation (NOP) logger implementation\n",
      "SLF4J: See https://www.slf4j.org/codes.html#noProviders for further details.\n",
      "SLF4J: Class path contains SLF4J bindings targeting slf4j-api versions 1.7.x or earlier.\n",
      "SLF4J: Ignoring binding found at [jar:file:/Users/kferrite/dev/gnomad_methods/venv/lib/python3.11/site-packages/pyspark/jars/log4j-slf4j-impl-2.17.2.jar!/org/slf4j/impl/StaticLoggerBinder.class]\n",
      "SLF4J: See https://www.slf4j.org/codes.html#ignoredBindings for an explanation.\n",
      "Setting default log level to \"WARN\".\n",
      "To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).\n",
      "Running on Apache Spark version 3.3.2\n",
      "SparkUI available at http://wm12f-58b.broadinstitute.org:4041\n",
      "Welcome to\n",
      "     __  __     <>__\n",
      "    / /_/ /__  __/ /\n",
      "   / __  / _ `/ / /\n",
      "  /_/ /_/\\_,_/_/_/   version 0.2.125-6e6f46797aed\n",
      "LOGGING: writing to /Users/kferrite/dev/gnomad_methods/notebooks/hail-20231128-1116-0.2.125-6e6f46797aed.log\n"
     ]
    }
   ],
   "source": [
    "# The clingen-public bucket will only be the week of 2023-09-18\n",
    "# The clingen-public-requesterpays bucket will remain available but the client will need to\n",
    "# be authenticated with a Google Cloud account with billing enabled to pay network transfer fees\n",
    "\n",
    "# configuration for data outputs\n",
    "bucket = \"clingen-public-requesterpays\"\n",
    "bucket = \"clingen-public\"\n",
    "\n",
    "# Writes inputs array as a hail table to this destination, if not None.\n",
    "# This can be useful for other testing, using this hail table as input without reconstructing it\n",
    "# inputs_ht_destination_url = f\"gs://{bucket}/gnomad-gks-qc/gnomad-filtered-inputs.ht\"\n",
    "inputs_ht_destination_url = None\n",
    "\n",
    "# Copies output annotations as newline delimited json to this url, if not None\n",
    "outputs_destination_file = f\"gs://{bucket}/gnomad-gks-qc/outputs.ndjson\"\n",
    "\n",
    "\n",
    "# ht_url can be a gs:// path, or a file:// local path\n",
    "\n",
    "### V3 tables\n",
    "# Publicly readable, but doesn't have all gnomad variants in it\n",
    "ht_url = f\"gs://{bucket}/gnomad-gks-downsampled-100k.ht\"\n",
    "\n",
    "# Can refer to a local hail table\n",
    "# ht_url = \"../downsample_to_100k_full_release.ht\"\n",
    "\n",
    "# ht_url = \"gs://gcp-public-data--gnomad/release/3.1.2/ht/genomes/gnomad.genomes.v3.1.2.sites.ht\"\n",
    "\n",
    "ht_url = \"gs://clingen-gnomad-mirror/gnomad.genomes.v3.1.2.sites_vrs.ht\"\n",
    "ht_version=\"3.1.4\"\n",
    "\n",
    "### V4 tables\n",
    "\n",
    "ht_url = \"gs://gcp-public-data--gnomad/release/4.0/ht/exomes/gnomad.exomes.v4.0.sites.ht\"\n",
    "# ht_url = \"gs://gcp-public-data--gnomad/release/4.0/ht/genomes/gnomad.genomes.v4.0.sites.ht/\"\n",
    "\n",
    "ht_version = \"4.0.0\"\n",
    "\n",
    "print(ht_url)\n",
    "ht = hl.read_table(ht_url)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5b26b21a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def loci_alleles_to_pandas(loci: List[hl.Locus], alleles: List[List[str]]) -> pd.DataFrame:\n",
    "    return pd.DataFrame({\n",
    "        \"locus\": loci,\n",
    "        \"alleles\": alleles\n",
    "    })\n",
    "\n",
    "def filter_table_to_loci_alleles(ht: hl.Table, loci: List[hl.Locus], alleles: List[List[str]]) -> hl.Table:\n",
    "    left = hl.Table.from_pandas(loci_alleles_to_pandas(loci, alleles))\n",
    "    left = left.key_by(left.locus, left.alleles)\n",
    "    return left.join(ht)\n",
    "\n",
    "def make_loci(inputs: List[Tuple[str,int]], reference_genome=\"GRCh38\") -> List[hl.Locus]:\n",
    "    return list(\n",
    "        map(lambda contig, pos: hl.locus(contig, pos, reference_genome=reference_genome),\n",
    "        map(lambda x: x[0], inputs),\n",
    "        map(lambda x: x[1], inputs)))\n",
    "\n",
    "# hl.Table.from_pandas(pd.DataFrame({\n",
    "#     \"locus\": [hl.locus(\"chr1\", 629844, reference_genome=\"GRCh38\")],\n",
    "#     \"alleles\": [[\"A\", \"G\"]]\n",
    "# })).key_by(\"locus\", \"alleles\").join(ht).show()\n",
    "\n",
    "# filter_table_to_loci_alleles(\n",
    "#     ht,\n",
    "#     make_loci([[\"chr1\", 629844]]),\n",
    "#     [[\"A\", \"G\"]]).show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "edac5b0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gnomad\n",
    "import gnomad.utils.annotations\n",
    "import gnomad.resources.grch38.gnomad\n",
    "\n",
    "# reload (re-running this cell will reload modifications to these modules on disk)\n",
    "import importlib\n",
    "importlib.reload(gnomad.utils.annotations)\n",
    "importlib.reload(gnomad.resources.grch38.gnomad)\n",
    "\n",
    "from gnomad.utils.annotations import add_gks_va, add_gks_vrs\n",
    "from gnomad.resources.grch38.gnomad import gnomad_gks\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "94145a04",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GnomAD 3.1.2\n",
    "# GRCh38 expressions\n",
    "\n",
    "inputs = [\n",
    "    # {\"gnomad\": \"1-55051215-G-GA\"},\n",
    "    # {\"gnomad\": \"1-629844-A-G\"},\n",
    "    # {\"gnomad\": \"1-633440-A-ATCCC\"},\n",
    "    # {\"gnomad\": \"1-37917308-TTATATA-T\"},\n",
    "    # {\"gnomad\": \"1-4807634-TAAAA-T\"},\n",
    "\n",
    "    # {\"gnomad\": \"X-152954317-T-C\"} # gnoamd v2\n",
    "    {\"gnomad\": \"X-152914462-G-C\"},\n",
    "    # {\"gnomad\": \"X-153688862-T-C\"}\n",
    "\n",
    "    {\"gnomad\": \"1-233194352-G-GT\"},\n",
    "    {\"gnomad\": \"1-95571386-C-A\"}\n",
    "]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f35b7333",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table><thead><tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td></tr><tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">locus</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">alleles</div></td></tr><tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">locus&lt;GRCh38&gt;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">array&lt;str&gt;</td></tr>\n",
       "</thead><tbody><tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">chr1:95571386</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">[&quot;C&quot;,&quot;A&quot;]</td></tr>\n",
       "<tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">chr1:233194352</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">[&quot;G&quot;,&quot;GT&quot;]</td></tr>\n",
       "<tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">chrX:152914462</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">[&quot;G&quot;,&quot;C&quot;]</td></tr>\n",
       "</tbody></table>"
      ],
      "text/plain": [
       "+----------------+------------+\n",
       "| locus          | alleles    |\n",
       "+----------------+------------+\n",
       "| locus<GRCh38>  | array<str> |\n",
       "+----------------+------------+\n",
       "| chr1:95571386  | [\"C\",\"A\"]  |\n",
       "| chr1:233194352 | [\"G\",\"GT\"] |\n",
       "| chrX:152914462 | [\"G\",\"C\"]  |\n",
       "+----------------+------------+"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# # This cell constructs an indexed hail table from the input alleles\n",
    "# # so the full gnomad dataset can be rapidly filtered by left joining it onto this.\n",
    "# # Using a .filter() to check membership in the input set is slow because it needs\n",
    "# # to do a table scan.\n",
    "\n",
    "input_gnomad_expressions = hl.literal([x[\"gnomad\"] for x in inputs])\n",
    "input_terms = [x[\"gnomad\"].split(\"-\") for x in inputs]\n",
    "\n",
    "df = pd.DataFrame(\n",
    "    {\n",
    "        \"contig\": [str(\"chr\" + i[0]) for i in input_terms],\n",
    "        \"position\": [int(i[1]) for i in input_terms],\n",
    "        \"ref\": [i[2] for i in input_terms],\n",
    "        \"alt\": [i[3] for i in input_terms]\n",
    "    }\n",
    ")\n",
    "input_ht = hl.Table.from_pandas(df)\n",
    "input_ht = (input_ht\n",
    "    .annotate(\n",
    "        locus=hl.locus(input_ht.contig, input_ht.position, reference_genome=\"GRCh38\"),\n",
    "        alleles=hl.array([input_ht.ref, input_ht.alt]))\n",
    "    .drop(\"contig\", \"position\", \"ref\", \"alt\")\n",
    "    .key_by(\"locus\", \"alleles\"))\n",
    "\n",
    "input_ht.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "79f65c41",
   "metadata": {},
   "outputs": [],
   "source": [
    "ht_filtered = ht\n",
    "\n",
    "# NOTE: To keep table filtered to only the set of input variants, keep this line uncommented\n",
    "# ht_filtered = input_ht.join(ht)\n",
    "\n",
    "\n",
    "ht_filtered = ht_filtered.annotate(\n",
    "    genomic_coordinates = hl.format(\"%s-%s-%s-%s\",\n",
    "        ht_filtered.locus.contig[3:], # Remove 'chr'\n",
    "        hl.str(ht_filtered.locus.position),\n",
    "        ht_filtered.alleles[0],\n",
    "        ht_filtered.alleles[1]\n",
    "    )\n",
    ")\n",
    "\n",
    "# # Write the filtered gnomad table to storage\n",
    "# if inputs_ht_destination_url:\n",
    "#     ht_filtered.write(inputs_ht_destination_url, overwrite=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "05a6b9b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parameters for gnomad_gks/get_gks\n",
    "ancestry_group_short_names = gnomad.resources.grch38.gnomad.POPS[\"v3\"]\n",
    "ancestry_groups_full_name_map = gnomad.sample_qc.ancestry.POP_NAMES\n",
    "gnomad_version_label = \"3.1.4\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2d6092b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gnomAD allele strings: ['1-11994-T-C', '1-12016-G-A', '1-12060-CTGGAG-C', '1-12074-T-C', '1-12102-G-A']\n"
     ]
    }
   ],
   "source": [
    "vrs_variants = []\n",
    "select_fields = {\n",
    "    \"freq\": ht_filtered.freq,\n",
    "    # \"grpMaxFAF95\": ht_filtered.grpMaxFAF95,\n",
    "    \"info\": ht_filtered.info,\n",
    "    \"genomic_coordinates\": ht_filtered.genomic_coordinates,\n",
    "    \"vrs\": ht_filtered.info.vrs\n",
    "}\n",
    "# if \"jointGrpMaxFAF95\" in ht_filtered.row:\n",
    "#     select_fields[\"jointGrpMaxFAF95\"] = ht_filtered.jointGrpMaxFAF95\n",
    "\n",
    "records = ht_filtered.select(**select_fields)\n",
    "records = records.take(5)\n",
    "\n",
    "variant_strs = [r.genomic_coordinates for r in records]\n",
    "loci = [\n",
    "    hl.locus(contig=str(\"chr\" + v0), pos=int(v1), reference_genome=\"GRCh38\")\n",
    "    for [v0, v1, *_] in\n",
    "    [v.split(\"-\") for v in variant_strs]\n",
    "]\n",
    "\n",
    "print(f\"gnomAD allele strings: {variant_strs}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "4f66c7db",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 2:>                                                          (0 + 1) / 1]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Annotated 1 records in 6.525036096572876 seconds (6.525036 sec/rec)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/kferrite/dev/gnomad_methods/venv/lib/python3.11/site-packages/python_jsonschema_objects/__init__.py:49: UserWarning:\n",
      "\n",
      "Schema version http://json-schema.org/draft-07/schema not recognized. Some keywords and features may not be supported.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# This cell takes one locus present in the input table, performs VRS and VA annotation and prints it.\n",
    "\n",
    "# For performance on larger datasets, the interval should be much larger, at least a few thousand.\n",
    "\n",
    "ivl_0 =hl.locus_interval(loci[0].contig, loci[0].position, loci[0].position+1, reference_genome=\"GRCh38\")\n",
    "ht_locus_0 = (hl.filter_intervals(ht_filtered, [ivl_0]))\n",
    "\n",
    "# ivl_1 =hl.locus_interval(loci[1].contig, loci[1].position, loci[1].position+1, reference_genome=\"GRCh38\")\n",
    "# ht_locus_1 = (hl.filter_intervals(ht_filtered, [ivl_1]))\n",
    "\n",
    "ivl_full_chr1 = hl.locus_interval(\"chr1\", 1, 248956422, reference_genome=\"GRCh38\")\n",
    "\n",
    "import time\n",
    "t0 = time.time()\n",
    "gks_annotations = gnomad_gks(\n",
    "    locus_interval=ivl_0,\n",
    "    custom_ht=ht_filtered,\n",
    "    version=ht_version,\n",
    "    data_type=\"genomes\",\n",
    "    by_ancestry_group=True,\n",
    "    by_sex=False,\n",
    "    vrs_only=False,\n",
    "    skip_checkpoint=True,\n",
    "    skip_coverage=False\n",
    ")\n",
    "t1 = time.time()\n",
    "l = len(gks_annotations)\n",
    "td = t1 - t0\n",
    "print(f\"Annotated {l} records in {td} seconds ({td/l:.6f} sec/rec)\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6a1b8451",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"locus\": {\n",
      "    \"contig\": \"chr1\",\n",
      "    \"position\": 11994,\n",
      "    \"reference_genome\": \"GRCh38\"\n",
      "  },\n",
      "  \"alleles\": [\n",
      "    \"T\",\n",
      "    \"C\"\n",
      "  ],\n",
      "  \"gks_vrs_variant\": {\n",
      "    \"_id\": \"ga4gh:VA.nHlWYJXgiuvXrLAxQsraJ6r-daZ8ELL-\",\n",
      "    \"type\": \"Allele\",\n",
      "    \"location\": {\n",
      "      \"type\": \"SequenceLocation\",\n",
      "      \"sequence_id\": \"ga4gh:SQ.Ya6Rs7DHhDeg7YaOSg1EoNi3U_nQ9SvO\",\n",
      "      \"interval\": {\n",
      "        \"start\": {\n",
      "          \"type\": \"Number\",\n",
      "          \"value\": 11993\n",
      "        },\n",
      "        \"end\": {\n",
      "          \"type\": \"Number\",\n",
      "          \"value\": 11994\n",
      "        },\n",
      "        \"type\": \"SequenceInterval\"\n",
      "      },\n",
      "      \"_id\": \"ga4gh:VSL.wuO74hQ9_6ny64lKFgzWuJehwSYZgD-W\"\n",
      "    },\n",
      "    \"state\": {\n",
      "      \"type\": \"LiteralSequenceExpression\",\n",
      "      \"sequence\": \"C\"\n",
      "    }\n",
      "  },\n",
      "  \"gks_va_freq\": {\n",
      "    \"id\": \"gnomAD-4.0.0-chr1-11994-T-C\",\n",
      "    \"type\": \"CohortAlleleFrequency\",\n",
      "    \"label\": \"Overall Cohort Allele Frequency for chr1-11994-T-C\",\n",
      "    \"derivedFrom\": {\n",
      "      \"id\": \"gnomAD4.0.0\",\n",
      "      \"type\": \"DataSet\",\n",
      "      \"label\": \"gnomAD v4.0.0\",\n",
      "      \"version\": \"4.0.0\"\n",
      "    },\n",
      "    \"focusAllele\": {\n",
      "      \"_id\": \"ga4gh:VA.nHlWYJXgiuvXrLAxQsraJ6r-daZ8ELL-\",\n",
      "      \"type\": \"Allele\",\n",
      "      \"location\": {\n",
      "        \"type\": \"SequenceLocation\",\n",
      "        \"sequence_id\": \"ga4gh:SQ.Ya6Rs7DHhDeg7YaOSg1EoNi3U_nQ9SvO\",\n",
      "        \"interval\": {\n",
      "          \"start\": {\n",
      "            \"type\": \"Number\",\n",
      "            \"value\": 11993\n",
      "          },\n",
      "          \"end\": {\n",
      "            \"type\": \"Number\",\n",
      "            \"value\": 11994\n",
      "          },\n",
      "          \"type\": \"SequenceInterval\"\n",
      "        },\n",
      "        \"_id\": \"ga4gh:VSL.wuO74hQ9_6ny64lKFgzWuJehwSYZgD-W\"\n",
      "      },\n",
      "      \"state\": {\n",
      "        \"type\": \"LiteralSequenceExpression\",\n",
      "        \"sequence\": \"C\"\n",
      "      }\n",
      "    },\n",
      "    \"focusAlleleCount\": 0,\n",
      "    \"locusAlleleCount\": 0,\n",
      "    \"alleleFrequency\": 0.0,\n",
      "    \"cohort\": {\n",
      "      \"id\": \"ALL\"\n",
      "    },\n",
      "    \"ancillaryResults\": {\n",
      "      \"homozygotes\": 0\n",
      "    },\n",
      "    \"qualityMeasures\": {\n",
      "      \"qcFilters\": [\n",
      "        \"AS_VQSR\",\n",
      "        \"AC0\"\n",
      "      ],\n",
      "      \"lowComplexityRegion\": false,\n",
      "      \"heterozygousSkewedAlleleCount\": 0,\n",
      "      \"meanDepth\": 0.11176870552858141,\n",
      "      \"monoallelic\": false\n",
      "    },\n",
      "    \"subcohortFrequency\": [\n",
      "      {\n",
      "        \"id\": \"chr1-11994-T-C.AFR\",\n",
      "        \"type\": \"CohortAlleleFrequency\",\n",
      "        \"label\": \"African/African-American Cohort Allele Frequency for chr1-11994-T-C\",\n",
      "        \"focusAllele\": \"#/focusAllele\",\n",
      "        \"focusAlleleCount\": 0,\n",
      "        \"locusAlleleCount\": 0,\n",
      "        \"alleleFrequency\": 0.0,\n",
      "        \"cohort\": {\n",
      "          \"id\": \"AFR\",\n",
      "          \"characteristics\": [\n",
      "            {\n",
      "              \"name\": \"genetic ancestry\",\n",
      "              \"value\": \"African/African-American\"\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        \"ancillaryResults\": {\n",
      "          \"homozygotes\": 0\n",
      "        }\n",
      "      },\n",
      "      {\n",
      "        \"id\": \"chr1-11994-T-C.AMR\",\n",
      "        \"type\": \"CohortAlleleFrequency\",\n",
      "        \"label\": \"Latino Cohort Allele Frequency for chr1-11994-T-C\",\n",
      "        \"focusAllele\": \"#/focusAllele\",\n",
      "        \"focusAlleleCount\": 0,\n",
      "        \"locusAlleleCount\": 0,\n",
      "        \"alleleFrequency\": 0.0,\n",
      "        \"cohort\": {\n",
      "          \"id\": \"AMR\",\n",
      "          \"characteristics\": [\n",
      "            {\n",
      "              \"name\": \"genetic ancestry\",\n",
      "              \"value\": \"Latino\"\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        \"ancillaryResults\": {\n",
      "          \"homozygotes\": 0\n",
      "        }\n",
      "      },\n",
      "      {\n",
      "        \"id\": \"chr1-11994-T-C.ASJ\",\n",
      "        \"type\": \"CohortAlleleFrequency\",\n",
      "        \"label\": \"Ashkenazi Jewish Cohort Allele Frequency for chr1-11994-T-C\",\n",
      "        \"focusAllele\": \"#/focusAllele\",\n",
      "        \"focusAlleleCount\": 0,\n",
      "        \"locusAlleleCount\": 0,\n",
      "        \"alleleFrequency\": 0.0,\n",
      "        \"cohort\": {\n",
      "          \"id\": \"ASJ\",\n",
      "          \"characteristics\": [\n",
      "            {\n",
      "              \"name\": \"genetic ancestry\",\n",
      "              \"value\": \"Ashkenazi Jewish\"\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        \"ancillaryResults\": {\n",
      "          \"homozygotes\": 0\n",
      "        }\n",
      "      },\n",
      "      {\n",
      "        \"id\": \"chr1-11994-T-C.EAS\",\n",
      "        \"type\": \"CohortAlleleFrequency\",\n",
      "        \"label\": \"East Asian Cohort Allele Frequency for chr1-11994-T-C\",\n",
      "        \"focusAllele\": \"#/focusAllele\",\n",
      "        \"focusAlleleCount\": 0,\n",
      "        \"locusAlleleCount\": 0,\n",
      "        \"alleleFrequency\": 0.0,\n",
      "        \"cohort\": {\n",
      "          \"id\": \"EAS\",\n",
      "          \"characteristics\": [\n",
      "            {\n",
      "              \"name\": \"genetic ancestry\",\n",
      "              \"value\": \"East Asian\"\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        \"ancillaryResults\": {\n",
      "          \"homozygotes\": 0\n",
      "        }\n",
      "      },\n",
      "      {\n",
      "        \"id\": \"chr1-11994-T-C.FIN\",\n",
      "        \"type\": \"CohortAlleleFrequency\",\n",
      "        \"label\": \"Finnish Cohort Allele Frequency for chr1-11994-T-C\",\n",
      "        \"focusAllele\": \"#/focusAllele\",\n",
      "        \"focusAlleleCount\": 0,\n",
      "        \"locusAlleleCount\": 0,\n",
      "        \"alleleFrequency\": 0.0,\n",
      "        \"cohort\": {\n",
      "          \"id\": \"FIN\",\n",
      "          \"characteristics\": [\n",
      "            {\n",
      "              \"name\": \"genetic ancestry\",\n",
      "              \"value\": \"Finnish\"\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        \"ancillaryResults\": {\n",
      "          \"homozygotes\": 0\n",
      "        }\n",
      "      },\n",
      "      {\n",
      "        \"id\": \"chr1-11994-T-C.MID\",\n",
      "        \"type\": \"CohortAlleleFrequency\",\n",
      "        \"label\": \"Middle Eastern Cohort Allele Frequency for chr1-11994-T-C\",\n",
      "        \"focusAllele\": \"#/focusAllele\",\n",
      "        \"focusAlleleCount\": 0,\n",
      "        \"locusAlleleCount\": 0,\n",
      "        \"alleleFrequency\": 0.0,\n",
      "        \"cohort\": {\n",
      "          \"id\": \"MID\",\n",
      "          \"characteristics\": [\n",
      "            {\n",
      "              \"name\": \"genetic ancestry\",\n",
      "              \"value\": \"Middle Eastern\"\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        \"ancillaryResults\": {\n",
      "          \"homozygotes\": 0\n",
      "        }\n",
      "      },\n",
      "      {\n",
      "        \"id\": \"chr1-11994-T-C.NFE\",\n",
      "        \"type\": \"CohortAlleleFrequency\",\n",
      "        \"label\": \"Non-Finnish European Cohort Allele Frequency for chr1-11994-T-C\",\n",
      "        \"focusAllele\": \"#/focusAllele\",\n",
      "        \"focusAlleleCount\": 0,\n",
      "        \"locusAlleleCount\": 0,\n",
      "        \"alleleFrequency\": 0.0,\n",
      "        \"cohort\": {\n",
      "          \"id\": \"NFE\",\n",
      "          \"characteristics\": [\n",
      "            {\n",
      "              \"name\": \"genetic ancestry\",\n",
      "              \"value\": \"Non-Finnish European\"\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        \"ancillaryResults\": {\n",
      "          \"homozygotes\": 0\n",
      "        }\n",
      "      },\n",
      "      {\n",
      "        \"id\": \"chr1-11994-T-C.REMAINING\",\n",
      "        \"type\": \"CohortAlleleFrequency\",\n",
      "        \"label\": \"Remaining Cohort Allele Frequency for chr1-11994-T-C\",\n",
      "        \"focusAllele\": \"#/focusAllele\",\n",
      "        \"focusAlleleCount\": 0,\n",
      "        \"locusAlleleCount\": 0,\n",
      "        \"alleleFrequency\": 0.0,\n",
      "        \"cohort\": {\n",
      "          \"id\": \"REMAINING\",\n",
      "          \"characteristics\": [\n",
      "            {\n",
      "              \"name\": \"genetic ancestry\",\n",
      "              \"value\": \"Remaining\"\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        \"ancillaryResults\": {\n",
      "          \"homozygotes\": 0\n",
      "        }\n",
      "      },\n",
      "      {\n",
      "        \"id\": \"chr1-11994-T-C.SAS\",\n",
      "        \"type\": \"CohortAlleleFrequency\",\n",
      "        \"label\": \"South Asian Cohort Allele Frequency for chr1-11994-T-C\",\n",
      "        \"focusAllele\": \"#/focusAllele\",\n",
      "        \"focusAlleleCount\": 0,\n",
      "        \"locusAlleleCount\": 0,\n",
      "        \"alleleFrequency\": 0.0,\n",
      "        \"cohort\": {\n",
      "          \"id\": \"SAS\",\n",
      "          \"characteristics\": [\n",
      "            {\n",
      "              \"name\": \"genetic ancestry\",\n",
      "              \"value\": \"South Asian\"\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        \"ancillaryResults\": {\n",
      "          \"homozygotes\": 0\n",
      "        }\n",
      "      }\n",
      "    ]\n",
      "  }\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "print(json.dumps(gks_annotations[0], indent=2))\n",
    "\n",
    "for ann in gks_annotations:\n",
    "    ann_id = ann[\"gks_va_freq\"][\"id\"]\n",
    "    with open(f\"{ann_id}.json\", \"w\") as f_out:\n",
    "        json.dump(ann, f_out, indent=4)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6cc1fe66",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # This cell does a larger annotation on a chromosome interval (e.g. first 20k bases of APC gene)\n",
    "\n",
    "# # APC: Chromosome: 5; NC_000005.10 (112707498..112846239)\n",
    "# apc_start = 112707498\n",
    "# apc_end = 112846239\n",
    "# larger_interval = hl.locus_interval(\n",
    "#     contig=\"chr5\",\n",
    "#     start=apc_start,\n",
    "#     end=apc_start + 20000,\n",
    "#     reference_genome=\"GRCh38\")\n",
    "\n",
    "# import time\n",
    "# t0 = time.time()\n",
    "# gks_annotations = gnomad_gks(\n",
    "#     locus_interval=larger_interval,\n",
    "#     custom_ht=ht_filtered,\n",
    "#     version=\"3.1.4\",\n",
    "#     data_type=\"genomes\",\n",
    "#     by_ancestry_group=True,\n",
    "#     by_sex=True,\n",
    "#     vrs_only=False,\n",
    "#     skip_checkpoint=True,\n",
    "#     skip_coverage=False\n",
    "# )\n",
    "# t1 = time.time()\n",
    "# l = len(gks_annotations)\n",
    "# td = t1 - t0\n",
    "# print(f\"Annotated {l} records in {td} seconds ({td/l:.6f} sec/rec)\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c38f2cb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(json.dumps(gks_annotations[2], indent=2))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "43cf10a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Load the jsonschema that can be used for validating objects\n",
    "# import subprocess\n",
    "# import shutil\n",
    "# import json\n",
    "# va_spec_clone = \"gnomad-gks-v1_va-spec\"\n",
    "# va_spec_branch = \"gk-pilot\"\n",
    "# shutil.rmtree(va_spec_clone, ignore_errors=True)\n",
    "# p = subprocess.run([\"git\", \"clone\", \"https://github.com/ga4gh/va-spec\", \"gnomad-gks-v1_va-spec\"],\n",
    "#                    check=True)\n",
    "# p = subprocess.run([\"bash\", \"-c\",\n",
    "#                     f\"cd {va_spec_clone} && git checkout {va_spec_branch}\"],\n",
    "#                    check=True)\n",
    "# with open(f\"{va_spec_clone}/schema/cohortAlleleFreq.json\") as f:\n",
    "#     schema = json.load(f)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "0cf009ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import json\n",
    "import requests\n",
    "import yaml\n",
    "import jsonschema\n",
    "\n",
    "def get_json_http(url):\n",
    "    r = requests.get(url)\n",
    "    if r.status_code != 200:\n",
    "        raise RuntimeError(f\"Request failed:\\n{r.status_code} {r.content}\")\n",
    "    return json.loads(r.content.decode(\"utf-8\"))\n",
    "\n",
    "# schema = get_json_http(\"https://raw.githubusercontent.com/ga4gh/va-spec/gk-pilot/schema/cohortAlleleFreq.json\")\n",
    "# schema = get_json_http(\"https://raw.githubusercontent.com/ga4gh/va-spec/eea47fd23cd57feac58466702ef3a9a4265ecfa1/schema/cohortAlleleFreq.json\")\n",
    "schema = get_json_http(\"https://raw.githubusercontent.com/ga4gh/va-spec/1.0-alpha/schema/cohortAlleleFreq.json\")\n",
    "\n",
    "# Local schema\n",
    "# with open(os.path.expanduser(\"~/dev/va-spec/schema/cohortAlleleFreq.yaml\")) as f:\n",
    "#     schema = yaml.safe_load(f)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "13d53b65",
   "metadata": {},
   "outputs": [],
   "source": [
    "for ann in gks_annotations:\n",
    "    va_annotation = ann[\"gks_va_freq\"]\n",
    "    jsonschema.validate(\n",
    "        instance=va_annotation,\n",
    "        schema=schema)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
