{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5b5ee5c4",
   "metadata": {},
   "source": [
    "This notebook uploads two outputs to Google Cloud Storage buckets.\n",
    "\n",
    "1) The inputs array converted to a vcf-style hail table.\n",
    "2) The next upload is a JSON file of the input variants converted to GA4GH GKS VA/VRS annotations using gnomAD data.\n",
    "\n",
    "It can read from a locally-staged hail table directory, or use one in GCS. The table must have a pre-computed `.info.VRS` field. This is computed using `tgg_methods` `vrs_annotation_batch.py`, which uses the `vrs-python` `vcf_annotation.py`\n",
    "\n",
    "https://github.com/broadinstitute/tgg_methods/blob/master/vrs/vrs_annotation_batch.py (last checked at `a0002f02fbd5dd25487b261e94081a3daec29c64`)\n",
    "\n",
    "This validates gnomad_methods functions `gnomad_gks` and `get_gks`, off the branch in this pull request: https://github.com/broadinstitute/gnomad_methods/pull/556\n",
    "\n",
    "The JSON schema is off the `aw-lb-drafting` branch (at `bf3b5aa`)\n",
    "\n",
    "The cell that patches the gnomad_gks function can be removed if the python environment this notebook is in is run off the version of code in the gnomad_methods PR 556."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e4985d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# configuration for data outputs\n",
    "\n",
    "# Writes inputs array as a hail table to this destination, if not None.\n",
    "# This can be useful for other testing, using this hail table as input without reconstructing it\n",
    "inputs_ht_destination_url = \"gs://clingen-public-requesterpays/gnomad-gks-larry-filtered.ht\"\n",
    "\n",
    "# Copies output annotations as newline delimited json to this url, if not None\n",
    "outputs_destination_file = \"gs://clingen-public-requesterpays/gnomad-gks-qc/outputs.ndjson\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "660728f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import hail as hl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edac5b0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gnomad\n",
    "import gnomad.utils.annotations\n",
    "import gnomad.resources.grch38.gnomad\n",
    "\n",
    "# reload (re-running this cell will reload modifications to these modules on disk)\n",
    "import importlib\n",
    "importlib.reload(gnomad.utils.annotations)\n",
    "importlib.reload(gnomad.resources.grch38.gnomad)\n",
    "\n",
    "from gnomad.utils.annotations import get_gks\n",
    "from gnomad.resources.grch38.gnomad import gnomad_gks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5d3a56c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "subprocess.run([\"rm\", \"-rf\", \"va-spec\"])\n",
    "p = subprocess.run([\"git\", \"clone\", \"https://github.com/ga4gh/va-spec\"],\n",
    "                   check=True)\n",
    "p = subprocess.run([\"bash\", \"-c\",\n",
    "                    \"cd va-spec && git checkout aw-lb-drafting\"],\n",
    "                   check=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf2e0aa5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "with open(\"va-spec/schema/cohortAlleleFreq.json\") as f:\n",
    "    schema = json.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94145a04",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GnomAD 3.1.2 \n",
    "# GRCh38 expressions\n",
    "\n",
    "inputs = [\n",
    "    {\n",
    "        \"clinvar\": \"422227\",\n",
    "        \"gnomad\": \"16-68737366-C-T\"\n",
    "    },\n",
    "    {\n",
    "        \"clinvar\": \"619\",\n",
    "        \"gnomad\": \"12-102843683-C-T\"\n",
    "    },\n",
    "    {\n",
    "       \"clinvar\": \"808527\",\n",
    "       \"gnomad\": \"19-38442434-C-T\"\n",
    "    },\n",
    "    {\n",
    "        \"clinvar\": \"695823\",\n",
    "        \"gnomad\": \"1-11130641-G-A\"\n",
    "    },\n",
    "    # Not in gnomAD 3.1.2\n",
    "#     {\n",
    "#         \"clinvar\": \"251162\",\n",
    "#         \"gnomad\": \"19-11105249-C-T\"\n",
    "#     },\n",
    "    {\n",
    "        \"clinvar\": \"143344\",\n",
    "        \"gnomad\": \"X-154030690-C-T\"\n",
    "    },\n",
    "    {\n",
    "        \"clinvar\": \"137390\",\n",
    "        \"gnomad\": \"14-28767768-C-T\"\n",
    "    },\n",
    "    {\n",
    "        \"clinvar\": \"129997\",\n",
    "        \"gnomad\": \"15-89317458-C-G\"\n",
    "    },\n",
    "    {\n",
    "        \"clinvar\": \"89198\",\n",
    "        \"gnomad\": \"2-47799491-C-G\"\n",
    "    },\n",
    "    {\n",
    "        \"clinvar\": \"43032\",\n",
    "        \"gnomad\": \"14-23416241-G-A\"\n",
    "    },\n",
    "    {\n",
    "        \"clinvar\": \"2356\",\n",
    "        \"gnomad\": \"1-216247118-C-A\"\n",
    "    },\n",
    "    {\n",
    "        \"clinvar\": \"3038\",\n",
    "        \"gnomad\": \"11-108248927-T-G\"\n",
    "    },\n",
    "    {\n",
    "        \"clinvar\": \"256155\",\n",
    "        \"gnomad\": \"7-107701083-T-A\"\n",
    "    },\n",
    "    # example variant from va-spec simple_result_example.yaml\n",
    "    {\n",
    "        \"gnomad\": \"1-55051215-G-GA\"\n",
    "    }\n",
    "]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "878db9fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ht_url can be a gs:// path, or a file:// local path\n",
    "\n",
    "# Publicly readable, but doesn't have all gnomad variants in it\n",
    "ht_url = \"gs://clingen-public-requesterpays/downsample_to_100k_full_release.ht\"\n",
    "\n",
    "# Can refer to a local hail table directory\n",
    "ht_url = \"../downsample_to_100k_full_release.ht\"\n",
    "\n",
    "# gnomad-vrs-io-finals is a private bucket, needs a service account with read access \n",
    "ht_url = \"gs://gnomad-vrs-io-finals/ht-outputs/release_0426_dmtest_v3.1.2-Full-ht-release-output-updated-schema-050523.ht\"\n",
    "\n",
    "ht = hl.read_table(ht_url)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f35b7333",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This cell constructs an indexed hail table from the input alleles\n",
    "# so the full gnomad dataset can be rapidly filtered by left joining it onto this.\n",
    "# Using a .filter() to check membership in the input set is slow because it needs\n",
    "# to do a table scan.\n",
    "# Too slow:\n",
    "# ht_with_coords.filter(input_gnomad_expressions.contains(ht_with_coords.genomic_coordinates))\n",
    "\n",
    "import pandas\n",
    "\n",
    "input_gnomad_expressions = hl.literal([x[\"gnomad\"] for x in inputs])\n",
    "input_terms = [x[\"gnomad\"].split(\"-\") for x in inputs]\n",
    "\n",
    "df = pandas.DataFrame(\n",
    "    {\n",
    "        \"contig\": [str(\"chr\" + i[0]) for i in input_terms],\n",
    "        \"position\": [int(i[1]) for i in input_terms],\n",
    "        \"ref\": [i[2] for i in input_terms],\n",
    "        \"alt\": [i[3] for i in input_terms]\n",
    "    }\n",
    ")\n",
    "inp_ht = hl.Table.from_pandas(df)\n",
    "inp_ht = (inp_ht\n",
    "    .annotate(\n",
    "        locus=hl.locus(inp_ht.contig, inp_ht.position, reference_genome=\"GRCh38\"),\n",
    "        alleles=hl.array([inp_ht.ref, inp_ht.alt]))\n",
    "    .drop(\"contig\", \"position\", \"ref\", \"alt\")\n",
    "    .key_by(\"locus\", \"alleles\"))\n",
    "\n",
    "inp_ht.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79f65c41",
   "metadata": {},
   "outputs": [],
   "source": [
    "ht_with_coords = ht.annotate(\n",
    "    genomic_coordinates = hl.format(\"%s-%s-%s-%s\",\n",
    "        ht.locus.contig[3:], # Remove 'chr'\n",
    "        hl.str(ht.locus.position),\n",
    "        ht.alleles[0],\n",
    "        ht.alleles[1]\n",
    "    )\n",
    ")\n",
    "\n",
    "# Indexed filter using join on the indexed vcf table constructed from inputs\n",
    "ht_filtered = inp_ht.join(ht_with_coords)\n",
    "ht_filtered.genomic_coordinates.show()\n",
    "\n",
    "if inputs_ht_destination_url:\n",
    "    ht_filtered.write(inputs_ht_destination_url, overwrite=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa595204",
   "metadata": {},
   "outputs": [],
   "source": [
    "# variant = ht_filtered.genomic_coordinates.take(1)[0]\n",
    "# print(variant)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aae718c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Patch gnomad_gks using code copied from local development environment\n",
    "\n",
    "# # need these references for monkey patch of gnomad_gks\n",
    "# from typing import Union\n",
    "# from gnomad.resources.grch38.gnomad import coverage, POPS\n",
    "# from gnomad.sample_qc.ancestry import POP_NAMES\n",
    "\n",
    "# def gnomad_gks(\n",
    "#     version: str,\n",
    "#     variant: str,\n",
    "#     data_type: str = \"genomes\",\n",
    "#     by_ancestry_group: bool = False,\n",
    "#     by_sex: bool = False,\n",
    "#     vrs_only: bool = False,\n",
    "#     ht: Union[str,hl.Table] = None,\n",
    "# ) -> dict:\n",
    "#     \"\"\"\n",
    "#     Call get_gks() and return VRS information and frequency information for the specified gnomAD release version and variant.\n",
    "\n",
    "#     :param version: String of version of gnomAD release to use.\n",
    "#     :param variant: String of variant to search for (chromosome, position, ref, and alt, separated by '-'). Example for a variant in build GRCh38: \"chr5-38258681-C-T\".\n",
    "#     :param data_type: String of either \"exomes\" or \"genomes\" for the type of reads that are desired.\n",
    "#     :param by_ancestry_group: Boolean to pass to obtain frequency information for each ancestry group in the desired gnomAD version.\n",
    "#     :param by_sex: Boolean to pass if want to return frequency information for each ancestry group split by chromosomal sex.\n",
    "#     :param vrs_only: Boolean to pass if only want VRS information returned (will not include allele frequency information).\n",
    "#     :param custom_ht_path: Path of Hail Table to parse if different from what the public_release() method would return for the version.\n",
    "#     :return: Dictionary containing VRS information (and frequency information split by ancestry groups and sex if desired) for the specified variant.\n",
    "\n",
    "#     \"\"\"\n",
    "#     # If ht is not already a table,\n",
    "#     # read in gnomAD release table to filter to chosen variant.\n",
    "#     if not isinstance(ht, hl.Table):\n",
    "#         if ht:\n",
    "#             ht = hl.read_table(ht)\n",
    "#         else:\n",
    "#             ht = hl.read_table(public_release(data_type).versions[version].path)\n",
    "\n",
    "#     high_level_version = f\"v{version.split('.')[0]}\"\n",
    "\n",
    "#     # Read coverage statistics.\n",
    "\n",
    "#     if high_level_version == \"v3\":\n",
    "#         coverage_version = \"3.0.1\"\n",
    "#     else:\n",
    "#         raise NotImplementedError(\n",
    "#             \"gnomad_gks() is currently only implemented for gnomAD v3.\"\n",
    "#         )\n",
    "\n",
    "#     coverage_ht = hl.read_table(coverage(data_type).versions[coverage_version].path)\n",
    "\n",
    "#     # Retrieve ancestry groups from the imported POPS dictionary.\n",
    "#     pops_list = list(POPS[high_level_version]) if by_ancestry_group else None\n",
    "\n",
    "#     # Throw warnings if contradictory arguments passed.\n",
    "#     if by_ancestry_group and vrs_only:\n",
    "#         logger.warning(\n",
    "#             \"Both 'vrs_only' and 'by_ancestry_groups' have been specified. Ignoring\"\n",
    "#             \" 'by_ancestry_groups' list and returning only VRS information.\"\n",
    "#         )\n",
    "#     elif by_sex and not by_ancestry_group:\n",
    "#         logger.warning(\n",
    "#             \"Splitting whole database by sex is not yet supported. If using 'by_sex',\"\n",
    "#             \" please also specify 'by_ancestry_group' to stratify by.\"\n",
    "#         )\n",
    "\n",
    "#     # Call and return get_gks() for chosen arguments.\n",
    "#     gks_info = get_gks(\n",
    "#         ht=ht,\n",
    "#         variant=variant,\n",
    "#         label_name=\"gnomAD\",\n",
    "#         label_version=version,\n",
    "#         coverage_ht=coverage_ht,\n",
    "#         ancestry_groups=pops_list,\n",
    "#         ancestry_groups_dict=POP_NAMES,\n",
    "#         by_sex=by_sex,\n",
    "#         vrs_only=vrs_only,\n",
    "#     )\n",
    "\n",
    "#     return gks_info\n",
    "# gnomad.resources.grch38.gnomad.gnomad_gks = gnomad_gks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05a6b9b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parameters for gnomad_gks/get_gks\n",
    "ancestry_group_short_names = gnomad.resources.grch38.gnomad.POPS[\"v3\"]\n",
    "ancestry_groups_full_name_map = gnomad.sample_qc.ancestry.POP_NAMES\n",
    "gnomad_version_label = \"3.1.4\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8fe3ca6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Test one variant in gnomad_gks\n",
    "# variant_gks = gnomad.resources.grch38.gnomad.gnomad_gks(\n",
    "#     version=\"3.1.4\",\n",
    "#     variant=\"chr1-55051215-G-GA\",\n",
    "#     data_type=\"genomes\",\n",
    "#     by_ancestry_group=True,\n",
    "#     by_sex=True,\n",
    "#     vrs_only=False,\n",
    "#     ht=ht_filtered\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89f4d888",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Test one variant in get_gks\n",
    "# variant_gks = get_gks(\n",
    "#     ht=ht_filtered,\n",
    "#     variant=\"chr1-55051215-G-GA\",\n",
    "#     label_name=\"gnomAD\",\n",
    "#     label_version=gnomad_version_label,\n",
    "#     coverage_ht=None,\n",
    "#     ancestry_groups=ancestry_group_short_names,\n",
    "#     ancestry_groups_dict=ancestry_groups_full_name_map,\n",
    "#     by_sex=False,\n",
    "#     vrs_only=False\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c279d9bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# json.dumps(variant_gks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36863043",
   "metadata": {},
   "outputs": [],
   "source": [
    "gks_annotations = []\n",
    "variant_strs = ht_filtered.genomic_coordinates.collect()\n",
    "print(variant_strs)\n",
    "for variant_str in variant_strs:\n",
    "    print(\"calling get_gks on: \" + variant_str)\n",
    "    va_freq = get_gks(\n",
    "        ht=ht_filtered,\n",
    "        variant=str(\"chr\" + variant_str),\n",
    "        label_name=\"gnomAD\",\n",
    "        label_version=gnomad_version_label,\n",
    "        coverage_ht=None,\n",
    "        ancestry_groups=ancestry_group_short_names,\n",
    "        ancestry_groups_dict=ancestry_groups_full_name_map,\n",
    "        by_sex=False,\n",
    "        vrs_only=False\n",
    "    )\n",
    "    gks_annotations.append(va_freq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a105c04",
   "metadata": {},
   "outputs": [],
   "source": [
    "import jsonschema\n",
    "for ann in gks_annotations:\n",
    "    jsonschema.validate(schema=schema, instance=ann)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41533fe6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import hailtop.fs as hlfs\n",
    "\n",
    "homedir = os.path.expanduser(\"~\")\n",
    "with open(homedir + \"/gnomad_gks_annotations.ndjson\", \"w\") as f:\n",
    "    for ann in gks_annotations:\n",
    "        json.dump(ann, f)\n",
    "        f.write(\"\\n\")\n",
    "\n",
    "if outputs_destination_file:\n",
    "    hlfs.copy(f\"file://{homedir}/gnomad_gks_annotations.ndjson\", outputs_destination_file)\n",
    "    print(\"Copied annotations to: \" + outputs_destination_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e798ceea",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}